test/evaluate/folding03.f90

```diff
@@ -0,0 +1,305 @@
+! Copyright (c) 2019, NVIDIA CORPORATION.  All rights reserved.
+!
+! Licensed under the Apache License, Version 2.0 (the "License");
+! you may not use this file except in compliance with the License.
+! You may obtain a copy of the License at
+!
+!     http://www.apache.org/licenses/LICENSE-2.0
+!
+! Unless required by applicable law or agreed to in writing, software
+! distributed under the License is distributed on an "AS IS" BASIS,
+! WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+! See the License for the specific language governing permissions and
+! limitations under the License.
+
+! Test operation folding edge case (both expected value and messages)
+! These tests make assumptions regarding real(4) and integer(4) extrema.
+
+#define TEST_ISNAN(v) logical, parameter :: test_##v =.NOT.(v.EQ.v)
+
+! TODO: 2 behaviors to clarify:
+
+!  Biggest negative literal integer(4): -(2**31) or -(2**31-1) ?
+!   -> Fortran standard says representable value set must be in [-HUGE, HUGE].
+!      For f18 integer(4) that means [-(2**31-1), 2**31-1].
+!   -> Current f18 behavior is to emit an error for -2147483648_4 literal (-2**31), but not for
+!      -2147483647_4 -1_4 which is folded to -2147483648_4 without errors/warnings.
+!  It seems to me that biggest negative literal and representable value should
+!  be the same.
+!   -> pgfortran and ifort are OK with -2147483648_4 literal.
+!   -> gfortran gives an error for both -2147483648_4 unless -fno-range-check is used.
+!      It never raises errors for -2147483647_4 -1_4 and fold it to -2147483648_4.
```

Comment:
As I noted in a comment above, `-2147483647_4` parses to a unary negation `-(2147483647_4)`.  If you want to allow it, you'll need to change the grammar.

Reply:
Thanks for explaining, I had not noticed that `literal-constant` was the union of the literal constants without signs. The funny thing is that `data i /-2147483648_4/ ` is then ok because from the Fortran standard says `signed-int-literal-constant` in R845.
Anyway, I do not think this edge case is worth modifying the grammar without knowing if it is actually useful.